File: FrontendAngular/src/app/department-list/shared/department.service.ts
import { Injectable } from "@angular/core";
import { IDepartment } from './department.interface';
import { Observable } from "rxjs";
import { HttpClient } from "@angular/common/http";

@Injectable()
export class DepartmentService {
    private readonly apiUrl: string = 'http://localhost:4200/rest/department';

    constructor(private httpClient: HttpClient) {
    }

    public addDepartment(department: IDepartment): Observable<null> {
        return this.httpClient.post<null>(`${this.apiUrl}/create`, department);
    }

    public deleteDepartment(name: string): Observable<object> {
        return this.httpClient.delete(`${this.apiUrl}/${name}/delete`);
    }

    public getDepartments(): Observable<IDepartment[]> {
        return this.httpClient.get<IDepartment[]>(this.apiUrl);
    }

    public updateDepartment(department: IDepartment): Observable<object> {
        return this.httpClient.put(`${this.apiUrl}/update`, department);
    }
}

File: WebApi/Services/DepartmentService.cs
using UniversityWebApi.Domain;
using UniversityWebApi.Dto;
using UniversityWebApi.Repositories;

namespace UniversityWebApi.Services
{
    public class DepartmentService : IDepartmentService
    {
        private readonly IDepartmentRepository _departmentRepository;

        public DepartmentService(IDepartmentRepository departmentRepository)
        {
            _departmentRepository = departmentRepository;
        }
        public int CreateDepartment(DepartmentDto department)
        {
            if (department == null)
            {
                throw new Exception($"{nameof(Department)} not found");
            }
            Department departmentEntity = department.ConvertToDepartment();
            return _departmentRepository.Create(departmentEntity);
        }

        public void DeleteDepartment(string name)
        {
            Department department = _departmentRepository.GetByName(name);
            if (department == null)
            {
                throw new Exception($"{nameof(Department)} not found, #Name - {name}");
            }
            _departmentRepository.Delete(department);
        }

        public Department GetDepartmentByName(string name)
        {
            Department department = _departmentRepository.GetByName(name);
            if (department == null)
            {
                throw new Exception($"{nameof(Department)} not found, #Name - {name}");
            }
            return department;
        }

        public List<Tuple<string, int>> GetDepartmentCountGroupByAddress()
        {
            return _departmentRepository.GetDepartmentCountGroupByAddress();
        }

        public List<Department> GetDepartments()
        {
            return _departmentRepository.GetAll();
        }

        public int UpdateDepartment(DepartmentDto department)
        {
            if (department == null)
            {
                throw new Exception($"{nameof(Department)} not found");
            }
            Department departmentEntity = department.ConvertToDepartment();
            Department departmentTemp = _departmentRepository.GetByName(departmentEntity.Name);
            if (departmentTemp == null)
            {
                throw new Exception($"{nameof(Department)} not found");
            }
            return _departmentRepository.Update(departmentEntity);
        }
    }
}
